File: morale-assistant-backend/src/main/java/hu/herolds/projects/moraleassistant/model/converter/URIAttributeConverter.java
package hu.herolds.projects.moraleassistant.model.converter;

import lombok.extern.slf4j.Slf4j;

import javax.persistence.AttributeConverter;
import javax.persistence.Converter;
import java.net.URI;
import java.net.URISyntaxException;

@Slf4j
@Converter(autoApply = true)
public class URIAttributeConverter implements AttributeConverter<URI, String> {

    @Override
    public String convertToDatabaseColumn(URI uri) {
        return uri != null ? uri.getPath() : null;
    }

    @Override
    public URI convertToEntityAttribute(String str) {
        try {
            return new URI(str);
        } catch (URISyntaxException e) {
            log.error("Could not convert to entity attribute (URI)", e);
            return null;
        }
    }
}

File: morale-assistant-backend-kt/src/main/kotlin/hu/herolds/projects/morale/domain/converter/URIAttributeConverter.kt
package hu.herolds.projects.morale.domain.converter

import org.slf4j.LoggerFactory
import java.net.URI
import java.net.URISyntaxException
import javax.persistence.AttributeConverter
import javax.persistence.Converter

@Converter(autoApply = true)
internal class URIAttributeConverter : AttributeConverter<URI?, String?> {
    private val log = LoggerFactory.getLogger(javaClass)

    override fun convertToDatabaseColumn(uri: URI?): String? {
        return uri?.path
    }

    override fun convertToEntityAttribute(str: String?): URI? {
        return try {
            str?.let {
                URI(it)
            }
        } catch (e: URISyntaxException) {
            log.error("Could not convert to entity attribute (URI)", e)
            null
        }
    }
}